### ISSUE TITLE
[Suggestion] Make boats be a seperate player model, not an entity.

### ISSUE THREAD

**Issue by C1ffisme** [minetest/minetest_game#830]
The current approach to boats is to use both an entity and a player, the player attached to the boat.

However, there is a less laggy approach, that lets minetest use only half as many objects.

Instead of attaching the player to the boat upon it being clicked, the boat is deleted and the player is teleported to the position, has his/her model changed, and then change their velocity based on boat programming (might require an engine change, I have seen that set_player_velocity or get_player_velocity do not exist yet.)

This also allows us to add other things, like carts or planes, without the lag that they usually create. (Although minecarts with items in them might be a problem.)

Not only this, but you could use no entities at all, simply changing the player to "boat mode" upon clicking a meshnode that looks like a boat.

I'm not sure if this is currently do-able because of the player_velocity thing, but I wanted to get the idea out there.

**Comment by xeranas** [comment]
What kind of hardware you guys use if extra block/entity is serious problem? I know games who are written in Java and did not have to go worry too much about extra block or two. :)

Frankly, this idea sounds like strange workaround. AFAIK attach/deattach is just api for that. If it creates lag or works works inefficiant - api should be fixed (if needed engine performance should be improved).  

> This also allows us to add other things, like carts or planes, without the lag that they usually create. 

I myself use carts mod and did not notice lag. Although it may create lag for some system because it checks in `on_step` tons of stuff (which direction it should go, how fast, if it still on rails and so on) - same stuff you would need to check if you would be one mesh instead of two. I think you looking into problem from wrong angle.

I would recommend fill a performance bug specific mod, which you may think it could perform better.

**Comment by C1ffisme** [comment]
@xeranas Entities are buggy in some ways. Maybe not on a non-laggy server, but if you get on a really laggy one, you could drop a mese crystal/diamond somewhere were it doesn't get stolen, wait a while, and then reap much more than you left.

This seems a bit more sensible than the current method, too. Rather than keeping two objects, both with the same position and yaw, we only have one that doesn't need to update the other's position, position, yaw, etc.

Although, I would like Minetest to get to the point where we don't have to worry about entity duplication, or the deletion of several entities because more than 49 are in 1 chunk. If we ever see this day, I would be happy. Then mobs and vehicles and large chunks of falling sand can co-exist at the same time, without crashing minetest or deleting themselves.

**Comment by xeranas** [comment]
@C1ffisme did you did some poc with profiling or just guessing? As I said you still need calculate same stuff. If you need move properly object you need apply rules how it should behave with environment those calculations are bottleneck, not extra object.

> Entities are buggy in some ways
Be more specific

You know you can work on Transformers mod. Sounds like fun idea and you can test your teory. If player on rails and press special key he transforms into cart. If he in air (jump) he transform into plane. Hmm? :)

Because maybe you right and I am wrong, but you need show some proof of concept to prove your point. "I think..." its weak argument. There lots of stuff in transportations mods where needs to be aware. Its not same thing as moving player.

**Comment by C1ffisme** [comment]
@xeranas Well, some of it I'm not sure of. I've taken some breaks from minetest every now and then, playing other things. Maybe some bugs have been fixed, but I wouldn't be the expert. I remember a day when entities in mods were deeply frowned upon because of the glitches in minetest that caused them to duplicate or disappear.

I'm not sure if these bugs still appear, though. They might have been fixed before I looked into getting a Github account to watch minetest's progress.
